installinator/src/peers.rs

```diff
@@ -0,0 +1,435 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{
+    fmt, future::Future, net::SocketAddrV6, pin::Pin, str::FromStr,
+    time::Duration,
+};
+
+use anyhow::{bail, Result};
+use buf_list::BufList;
+use bytes::Bytes;
+use display_error_chain::DisplayErrorChain;
+use futures::StreamExt;
+use itertools::Itertools;
+use tokio::{sync::mpsc, time::Instant};
+
+use crate::{
+    ddm_admin_client::DdmAdminClient,
+    errors::{ArtifactFetchError, DiscoverPeersError},
+};
+
+/// A chosen discovery mechanism for peers, passed in over the command line.
+#[derive(Clone, Debug)]
+pub(crate) enum DiscoveryMechanism {
+    /// The default discovery mechanism: hit the bootstrap network.
+    Bootstrap,
+
+    /// A list of peers is manually specified.
+    List(Vec<SocketAddrV6>),
+}
+
+// TODO: This currently hardcodes this wicketd port, will probably want to sync up on this.
+const WICKETD_PORT: u16 = 14000;
```

Comment:
@andrewjstone I've currently hardcoded the wicketd port here.

One thing I'm wondering is if we should have a separate dropshot server for serving artifacts, that both wicketd and sled-agent both spin up. sled-agent doesn't currently have an HTTP server exposed on the bootstrap network at all, so this "artifact server" could be shared by both services. Installinator wouldn't need to know about wicketd vs sled-agent: it could just hit the artifact port.

Reply:
@sunshowers I think that makes a lot of sense. Let's do it!

Also, great work on this whole PR and all the stuff you got working today. I'm sorry I haven't had a chance to dig in deeper on this this week. From my cursory looks periodically everything appears to be going swimmingly :)